hpx/lcos/detail/async_implementations.hpp

```diff
@@ -42,42 +36,70 @@ namespace hpx { namespace detail
         naming::id_type gid_;
     };
 
-    ///////////////////////////////////////////////////////////////////////////
-    template <typename Action, typename Result>
-    struct sync_local_invoke
+    template <typename Result>
+    class handle_managed_target
     {
-        template <typename ...Ts>
-        static lcos::future<Result>
-        call(naming::id_type const& id, naming::address && addr, Ts &&... vs)
-        {
-            bool target_is_managed = false;
-            naming::id_type id1;
+        HPX_NON_COPYABLE(handle_managed_target);
 
+    public:
+        handle_managed_target(hpx::id_type const& id, future<Result>& f)
+          : target_is_managed_(false), f_(f)
+        {
             if (id.get_management_type() == naming::id_type::managed)
             {
-                id1 = naming::id_type(id.get_gid(), naming::id_type::unmanaged);
-                target_is_managed = true;
+                id_ = naming::id_type(id.get_gid(), naming::id_type::unmanaged);
+                target_is_managed_ = true;
             }
+            else
+            {
+                id_ = id;
```

Comment:
A little, yes. But you are right ... the fact that we take the id_type by const reference is certainly a bit unfortunate here.